include/sys/mutex.h

```diff
@@ -176,19 +178,22 @@ spl_mutex_lockdep_on_maybe(kmutex_t *mp)			\
  *
  * However, there are many places in ZFS where the mutex is used for
  * serializing object freeing, and the code is shared among other OSes without
- * this issue. Thus, we need the spinlock to force the serialization on
- * mutex_exit().
+ * this issue. Thus, we need spin in mutex_destroy() until we're sure that all
+ * threads have finished slow-path in mutex_exit(), avoiding use-after-free.
  *
  * See http://lwn.net/Articles/575477/ for the information about the race.
  */
-#define	mutex_exit(mp)						\
-{								\
-	spl_mutex_lockdep_off_maybe(mp);			\
-	spin_lock(&(mp)->m_lock);				\
-	spl_mutex_clear_owner(mp);				\
-	mutex_unlock(MUTEX(mp));				\
-	spin_unlock(&(mp)->m_lock);				\
-	spl_mutex_lockdep_on_maybe(mp);				\
+
+#undef mutex_destroy
+#define	mutex_destroy(mp)						\
+{									\
+	for(;;)	{							\
+		VERIFY3P(mutex_owner(mp), ==, NULL);			\
```

Comment:
Fixed. Hitting that VERIFY anywhere in `mutex_destroy()` would be a bug (destroying a locked mutex). Removal of the spinlock does not change semantics of this condition, the `owner` is protected by the real mutex.